### PR TITLE
ocr v1

### LINKED ISSUE #13: first implementation of the OCR
I was thinking of trying this OCR for V1:
https://github.com/kha-white/manga-ocr

It only works for japanese, but I think it's a good place to start. 
A significant portion of our dataset is in Japanese.

### PR DESCRIPTION
**Pull request by PladsElsker**
Suggested implementation for v1 with: 
- manga-ocr
- ocr strategy pattern

closes #13

### CODE REVIEW

**Comment by PladsElsker** [comment]
Is there a way to keep the already installed version of torch+cuda when adding libraries that need torch? 
Right now, I think I have to run again `poe torch-cuda` to make sure I'm using the cuda version, and installing the env with `poetry install` seems to install an additional cpu version of pytorch, which is inconvenient. 

Any help would be appreciated!

File: manga_auto_translator/main.py [hunk 1]
```diff
@@ -16,10 +17,16 @@
...
+    '--ocr', 
+    type=click.Choice(ALLOWED_OCR_OPTIONS),
+    default='manga-ocr',
+    help='OCR strategy to use for character recognition.'
+)
+def cli(input_path, output_path, **kwargs):
     scans = ScanIOManager.load_scans(input_path)
 
-    pipeline = TranslationPipeline(scans)
+    pipeline = TranslationPipeline(scans, kwargs)
```
Comment (lbaret) [inline manga_auto_translator/main.py]: Generally we use **kwargs

File: manga_auto_translator/ocr.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+class OcrStrategyFactory:
+    def __init__(self, strategy: str) -> None:
+        self.selected = strategy
+
+    def create(self) -> OcrStrategy:
+        return available_strategies[self.selected]()
+
+
+available_strategies = {
```
Comment (lbaret) [inline manga_auto_translator/ocr.py]: At my company we use a children custom module inherited from Enum to make clean conversions.
https://docs.python.org/3/library/enum.html

File: manga_auto_translator/pipeline.py [hunk 3]
```diff
@@ -1,10 +1,12 @@
 from typing import Sequence
 from manga_auto_translator.data_structure import Scan
+from manga_auto_translator.ocr import OcrStrategyFactory
 
 
 class TranslationPipeline:
-    def __init__(self, scans: Sequence[Scan]) -> None:
+    def __init__(self, scans: Sequence[Scan], pipeline_args: dict) -> None:
```
Comment (lbaret) [inline manga_auto_translator/pipeline.py]: FYI : Dict from typing allow you to make a clean typing with python dictionnary. The format is the following :

```py
from typing import Dict

CUSTOM_TYPE = Dict[<type of key>, <type of value>]
```

File: manga_auto_translator/pipeline.py [hunk 4]
```diff
@@ -1,10 +1,12 @@
 from typing import Sequence
 from manga_auto_translator.data_structure import Scan
+from manga_auto_translator.ocr import OcrStrategyFactory
 
 
 class TranslationPipeline:
-    def __init__(self, scans: Sequence[Scan]) -> None:
+    def __init__(self, scans: Sequence[Scan], pipeline_args: dict) -> None:
         self.scans = scans
+        self.ocr_strategy = OcrStrategyFactory(strategy=pipeline_args.get('ocr')).create()
```
Comment (lbaret) [inline manga_auto_translator/pipeline.py]: It's better to instanciate modules outside of the module scope and pass it through argument.
This way we can't properly test the module.

**Review by lbaret — APPROVED** [review]

**Comment by lbaret** [comment]
> Is there a way to keep the already installed version of torch+cuda when adding libraries that need torch? Right now, I think I have to run again `poe torch-cuda` to make sure I'm using the cuda version, and installing the env with `poetry install` seems to install an additional cpu version of pytorch, which is inconvenient.
> 
> Any help would be appreciated!

I often have the same problem. I think it come from pip itself trying to match actual package requirements ...

File: manga_auto_translator/pipeline.py (hunk 4 above)
Comment (PladsElsker) [reply manga_auto_translator/pipeline.py]: ~~The problem is that it takes forever to load the module, so I don't think we want to load it if we're not using it (so if we pass anything else than `--ocr manga-ocr`).~~

Comment (PladsElsker) [reply manga_auto_translator/pipeline.py]: Oh nvm misread that I think, all good I'll do that

**Pull request merged**